### ISSUE TITLE
Allow specifying `serde` path and avoid injecting `extern crate serde as _serde`.

### ISSUE THREAD

**Issue by carllerche** [serde-rs/serde#1487]
[tower-web](https://github.com/carllerche/tower-web/) provides a `#[derive(Response)]` macro. This macro is a shim on top of serde's `derive(Serialize)`. It works by catching HTTP specific attributes and leaving any serde attributes. It then generates a "shadow" version of the struct that has the actual `derive(Serialize)`.

The problem is that the user must add `serde` to their `Cargo.toml` even if they do not interact with any serde APIs directly.

`tower-web` re-exports `serde` in the crate. I propose to have a serde attribute that allows specifying the path to `serde`. If this attribute is set, the `extern crate serde as _serde` is omitted.

[This](https://gist.github.com/carllerche/6e2351e1c3f968792cb898c08a726d96) gist illustrates the problem. `hello.rs` is very basic. I also included the expanded code.

In short, I would like [this](https://gist.github.com/carllerche/6e2351e1c3f968792cb898c08a726d96#file-hello-expanded-rs-L70) line omitted and all references to `_serde` replaced with `__tw::codegen::serde`.

One way to expose the API is as an attribute:

```rust
#[derive(Serialize)]
#[serde(serde_path = "__tw::codegen::serde")]
struct Routes {
// ....
}
```

If you like this strategy, I could attempt a PR.

**Comment by dtolnay** [comment]
Seems reasonable. Please send a PR.

You will need to decide whether the given path is a `use` path or a relative path i.e. whether the expanded code is:

```rust
use __tw::codegen::serde;
impl serde::Serialize for T
```

or this:

```rust
impl __tw::codegen::serde::Serialize for T
```

Those would have different behavior in 2015 edition.

**Comment by carllerche** [comment]
I'm pretty sure `use` doesn't work in the anon const scope thing. Since tower-web generates in a const scope, it cannot do option 1.

My thought is that whatever is set in `#[serde(serde_path = "__tw::codegen::serde")]` is used verbatim in the generated code. i.e. option 2.

Do you have thoughts on this? Also, do you have thoughts on what the attr should be?

**Comment by carllerche** [comment]
Gentle ping :)

When you are +1 on the path, I can try to put together a PR (since it will probably be non-trivial, I'm hoping to get approval on the direction first).

**Comment by dtolnay** [comment]
Thanks for the reminder. I would accept a PR for option 2 using:

```rust
#[serde(crate = "__tw::codegen::serde")]
```

as the attribute.
